### PR TITLE
Initial GPU support

### PR DESCRIPTION
**Pull request by atmyers**
This PR adds initial support for GPU offloading to WarpX. Along with changes made in PICSAR and AMReX, this pull request allows the complete PIC loop to run on multiple GPUs on the uniform plasma test problem for several standard algorithmic options you can set in your inputs file. 

Portions of the code that are offloaded include:
    EvolveParticles (provided the choices of current deposition, particle pushing, and field gathering are those offloaded in PICSAR)
    The finite difference version of pushing E and B (ditto)
    The laser particle update
    
There is additionally a fair bit of refactoring involved, both to minimize CPU / GPU page faults, and to reduce code duplication between the laser and physical particle updates.

Finally, an option to sort the particles every N time steps on the GPU was added, which is found to reduce the average time per step over long time evolutions.

### CODE REVIEW

**Comment by RemiLehe** [comment]
Thanks! This is a minor point but: should we remove the `USE_CUDA=FALSE` in the `WarpX-tests.ini`? (Now that the default compiler flags have been reverted to the CPU ones.)

File: Source/Make.WarpX [hunk 1]
```diff
@@ -5,6 +5,13 @@ OPENBC_HOME ?= ../../../openbc_poisson
 USE_MPI       = TRUE
 USE_PARTICLES = TRUE
 
+ifeq ($(USE_GPU),TRUE)
+  USE_OMP  = FALSE
+  USE_CUDA = TRUE
+  USE_ACC  = TRUE
+  NVCC_HOST_COMP = gcc
```
Comment (RemiLehe) [inline Source/Make.WarpX]: Out of curiosity, why do we need the above line?
(`NVCC_HOST_COMP = gcc`)
Is still being used even when we use the `pgi` compiler?

**Comment by RemiLehe** [comment]
Also, could you give a quick summary of the differences between `AddPlasmaCPU` and `AddPlasmaGPU`?

File: Source/Make.WarpX (hunk 1 above)
Comment (atmyers) [reply Source/Make.WarpX]: Yes, we actually use gcc to compile the host code, even when we use pgi to compile the device code. This was to work around limitations pgi had with some C++11 features we use, the details of which I'm not remembering right now.

**Comment by atmyers** [comment]
Right now the only difference between AddPlasmaGPU and AddPlasmaCPU is that the GPU version adds particles into a buffer on the host and then copies them all at once onto the device. But, I separated them into two functions in anticipation of later changes to the GPU version for when we do injection on the GPU.

**Pull request merged**